MantidPlot/src/ProjectRecovery.cpp

```diff
@@ -0,0 +1,354 @@
+
+    if (foundPath.isDirectory()) {
+      folderPaths.push_back(std::move(foundPath));
+    }
+    ++dirIterator;
+  }
+
+  size_t numberOfDirsPresent = folderPaths.size();
+  if (numberOfDirsPresent <= checkpointsToKeep) {
+    // Nothing to do
+    return;
+  }
+
+  // Ensure the oldest is first in the vector
+  std::sort(folderPaths.begin(), folderPaths.end(),
+            [](const Poco::Path &a, const Poco::Path &b) {
+              return a.toString() < b.toString();
+            });
+
+  size_t checkpointsToRemove = numberOfDirsPresent - checkpointsToKeep;
+  bool recurse = true;
+  for (size_t i = 0; i < checkpointsToRemove; i++) {
+    Poco::File(folderPaths[i]).remove(recurse);
+  }
+}
+
+/// Starts a background thread which saves out the project periodically
+void ProjectRecovery::startProjectSaving() {
+  // Close the existing thread first
+  stopProjectSaving();
+
+  if (!SAVING_ENABLED) {
+    return;
+  }
+
+  // Spin up a new thread
+  {
+    std::lock_guard<std::mutex> lock(m_notifierMutex);
+    m_stopBackgroundThread = false;
+  }
+
+  m_backgroundSavingThread = createBackgroundThread();
+}
+
+/// Stops any existing background threads which are running
+void ProjectRecovery::stopProjectSaving() {
+  {
+    std::lock_guard<std::mutex> lock(m_notifierMutex);
+    m_stopBackgroundThread = true;
+    m_threadNotifier.notify_all();
+  }
+
+  if (m_backgroundSavingThread.joinable()) {
+    m_backgroundSavingThread.join();
+  }
+}
+
+/// Top level thread wrapper which catches all exceptions to gracefully handle
+/// them
+void ProjectRecovery::projectSavingThreadWrapper() {
+  try {
+    projectSavingThread();
+  } catch (std::exception const &e) {
+    std::string preamble("Project recovery has stopped. Please report"
+                         " this to the development team.\nException:\n");
+    g_log.warning(preamble + e.what());
+  } catch (...) {
+    g_log.warning("Project recovery has stopped. Please report"
+                  " this to the development team.");
+  }
+}
+
+/**
+ * Main thread body which is run to save out projects. A member mutex is
+ * locked and monitored on a timeout to indicate if the thread should
+ * exit early. After the timeout elapses, if the thread has not been
+ * requested to exit, it will save the project out
+ */
+void ProjectRecovery::projectSavingThread() {
+  while (!m_stopBackgroundThread) {
+    std::unique_lock<std::mutex> lock(m_notifierMutex);
+    // The condition variable releases the lock until the var changes
+    if (m_threadNotifier.wait_for(lock, TIME_BETWEEN_SAVING, [this]() {
+          return m_stopBackgroundThread;
+        })) {
+      // Exit thread
+      g_log.debug("Project Recovery: Stopping background saving thread");
+      return;
+    }
+
+    g_log.debug("Project Recovery: Saving started");
+    // "Timeout" - Save out again
+    // Generate output paths
+    const auto basePath = getOutputPath();
+
+    Poco::File(basePath).createDirectory();
```

Comment:
On my OS, Ubuntu 16.04, I get an exception in the log after the first time project recovery saving kicks in:
```
Project recovery has stopped. Please report this to the development team.
Exception:
File not found
```
I think the issue is that the on linux, by default, when you create a directory the parent directories must exist but in my case the `$HOME/.mantid/recovery` doesn't exist so it cannot create the timestamped child directories.
I tested this by changing this line to `Poco::File(basePath).createDirectories()` and this seems to work.